src/formats/collada/ColladaScene.js

```diff
@@ -667,223 +816,106 @@ define([
 
-            this.transformationMatrix.multiplyByRotation(1, 0, 0, this.xRotation);
-            this.transformationMatrix.multiplyByRotation(0, 1, 0, this.yRotation);
-            this.transformationMatrix.multiplyByRotation(0, 0, 1, this.zRotation);
+            this._transformationMatrix.multiplyByRotation(1, 0, 0, this._xRotation);
+            this._transformationMatrix.multiplyByRotation(0, 1, 0, this._yRotation);
+            this._transformationMatrix.multiplyByRotation(0, 0, 1, this._zRotation);
 
-            this.transformationMatrix.multiplyByScale(this.scale, this.scale, this.scale);
+            this._transformationMatrix.multiplyByScale(this._scale, this._scale, this._scale);
 
-            this.transformationMatrix.multiplyByTranslation(this.xTranslation, this.yTranslation, this.zTranslation);
+            this._transformationMatrix.multiplyByTranslation(this._xTranslation, this._yTranslation, this._zTranslation);
 
             this.computeNormalMatrix();
-
         };
 
         // Internal. Intentionally not documented.
         ColladaScene.prototype.computeNormalMatrix = function () {
-
-            var rotAngles = new Vec3(0, 0, 0);
-
-            this.transformationMatrix.extractRotationAngles(rotAngles);
-
-            this.normalMatrix = Matrix.fromIdentity();
-
-            this.normalMatrix.multiplyByRotation(-1, 0, 0, rotAngles[0]);
-            this.normalMatrix.multiplyByRotation(0, -1, 0, rotAngles[1]);
-            this.normalMatrix.multiplyByRotation(0, 0, -1, rotAngles[2]);
+            this._transformationMatrix.extractRotationAngles(this._scratchVector);
+            this._normalTransformMatrix.setToIdentity();
+            this._normalTransformMatrix.multiplyByRotation(-1, 0, 0, this._scratchVector[0]);
+            this._normalTransformMatrix.multiplyByRotation(0, -1, 0, this._scratchVector[1]);
+            this._normalTransformMatrix.multiplyByRotation(0, 0, -1, this._scratchVector[2]);
         };
 
         // Internal. Intentionally not documented.
         ColladaScene.prototype.mustRenderNode = function (nodeId) {
             var draw = true;
-            if (this.hideNodes) {
-                var pos = this.nodesToHide.indexOf(nodeId);
+            if (this._hideNodes) {
+                var pos = this._nodesToHide.indexOf(nodeId);
                 draw = (pos === -1);
             }
             return draw;
         };
 
+        // Internal. Intentionally not documented.
+        ColladaScene.prototype.initializeUintExtension = function (dc) {
+            var ext = dc.gpuResourceCache.resourceForKey(ColladaScene._uintExtCacheKey);
+
+            if (ext) {
+                this._hasUintExt = true;
+            }
+            else {
+                this._hasUintExt = false;
+                ext = dc.currentGlContext.getExtension('OES_element_index_uint');
+                if (ext) {
+                    dc.gpuResourceCache.putResource(ColladaScene._uintExtCacheKey, ext, 1);
```

Comment:
Couldn't the extension be saved in a member variable of the class? Instead of the _hasUintExt, we could then check whether this new member variable is null or not for example.

What is the advantage of using the GPU resource cache?

Reply:
It will not work in a multi window setup.
GPU resources can not be shared between contexts, so only the first window will render the model, the rest will throw a webGL error.

However this approach has a flaw that the extension may be evicted from the cache and then we will have a conflict of state, where the instance says it’s true and the gpuResourceCache says it’s false.

I propose we don’t keep any information about the extension on the instance and just use the gpuResourceCache for now.

Alternatively we may look into a better approach about managing gl extensions.
For example, we can use a dedicated class or simply cache them on the drawContext and use a get method ```dc.getExtension('OES_element_index_uint')```

Reply:
Dave and I discussed this on Tuesday. The right place to store this is indeed as part of the DrawContext. It also detects context loses and can clear the cache accordingly. Could you propose the required changes to the DrawContext in a separate pull request? Thanks.